### ISSUE TITLE
Expose startWithNext/observeNext only for Error == NoError

### ISSUE THREAD

**Issue by NachoSoto** [ReactiveCocoa/ReactiveCocoa#3010]
I was thinking about how easy it is to do something like:
```swift
let signal: Signal<Int, SomeError>

signal.observeNext { print($0) }
```

And totally ignore errors. `RxJava` deals with this by throwing an exception at _runtime_ if this happens, but we can do better! We could detect this at _compile time_ if `observeNext`/`startWithNext` were not exposed for any `Signal`/`SignalProducer` that has an error other than `NoError`.

Obviously we can't know what users are doing inside of `observe`/`start`, but at least we can limit the damage.

I was going to suggest having `startWithNext:failed:`/`observeNext:failed:` in the case that the error **is not** `NoError`, but unfortunately that's not doable. I filed [`rdar://23432210`](http://www.openradar.me/23432210) a while ago asking for something that would support this though.
We could always add this overload though, it's just silly for signals with `NoError`.

So to summarize, I'm asking that we _hide_ `startWithNext`/`observeNext` behind:
```swift
extension SignalType where Error == NoError
extension SignalProducerType where Error == NoError
```

**Comment by andersio** [comment]
I think it is a logical change. Maybe we could use `observeResult` and `startWithResult` as the unconstrainted replacement.

**Comment by mdiep** [comment]
👍 That makes sense to me.

> I would suggest adding observeResult and startWithResult as the unconstrainted replacements if we adopt this proposal.

I like it. 👍

**Comment by NachoSoto** [comment]
I'm working on this now!

**Comment by NachoSoto** [comment]
I've kept the original to help people transition:
<img width="397" alt="screen shot 2016-06-28 at 10 23 24" src="https://cloud.githubusercontent.com/assets/685609/16425611/5e4cdf2c-3d1a-11e6-8ce3-78a067d5692e.png">
